src/ts/Query.ts

```diff
@@ -0,0 +1,305 @@
+}
+
+export interface FilterOperators<TValue> {
+	/** Equal */
+	$eq?: TValue;
+	/** Greater than */
+	$gt?: TValue;
+	/** Greater than or equal to */
+	$gte?: TValue;
+	/** Less than */
+	$lt?: TValue;
+	/** Less than or equal to */
+	$lte?: TValue;
+	/** Not equal */
+	$ne?: TValue;
+	/** String containing */
+	$contains?: TValue;
+	/** String starts with */
+	$startsWith?: TValue;
+	/** String ends with */
+	$endsWith?: TValue;
+	/** In list */
+	$in?: TValue[];
+	/** Not in list */
+	$nin?: TValue[];
+}
+
+export interface RootFilterOperators<TSchema> {
+	/** Used to combine conditions with an and */
+	$and?: Filter<TSchema>[];
+	/** Used to combine conditions with an or */
+	$or?: Filter<TSchema>[];
+}
+
+export type Condition<TValue> = TValue | TValue[] | FilterOperators<TValue>;
+
+export type Filter<TSchema extends GenericObject = GenericObject> = {
+	[key in keyof TSchema]?: Condition<TSchema[key]>;
+} & RootFilterOperators<TSchema>;
+
+export type SortCriteria = [string, -1 | 1][];
+
+export interface QueryExecutionResult {
+	/** Number of documents returned */
+	count: number;
+	/** Unformatted documents returned from database */
+	documents: DbDocument[];
+}
+// #endregion
+
+/** A query object */
+class Query<TSchema extends GenericObject = GenericObject> {
+	/** Model constructor to use with query */
+	private Model: ModelConstructor;
+
+	/** String to use as selection criteria in query */
+	private selection: string | null;
+
+	/** String to use as sort criteria in query */
+	private sort: string | null;
+
+	/** Limit the result set to this number of items */
+	private limit?: number;
+
+	/** Skip this number of items in the result set */
+	private skip?: number;
+
+	/** Specify the projection attribute in result set */
+	private projection: string[];
+
+	public constructor(
+		Model: ModelConstructor,
+		selectionCriteria: Filter<TSchema>,
+		options: QueryConstructorOptions = {},
+	) {
+		const { sort, limit, skip, projection } = options;
+
+		this.Model = Model;
+		this.limit = limit;
+		this.skip = skip;
+		this.projection = projection ?? [];
+
+		this.selection = this.formatSelectionCriteria(selectionCriteria);
+		this.sort = this.formatSortCriteria(sort);
+	}
+
+	/** Execute query */
+	public async exec(): Promise<QueryExecutionResult> {
+		let queryCommand = `select ${this.Model.file}`;
+		if (this.selection != null) {
+			queryCommand = `${queryCommand} with ${this.selection}`;
+		}
+		if (this.sort != null) {
+			queryCommand = `${queryCommand} ${this.sort}`;
+		}
+
+		const projection = this.Model.schema?.transformPathsToDbPositions(this.projection) ?? [];
+
+		const options = {
+			filename: this.Model.file,
+			queryCommand,
+			...(this.skip != null && { skip: this.skip }),
+			...(this.limit != null && { limit: this.limit }),
+			projection,
+		};
+
+		this.Model.connection.logger.debug(`executing query "${queryCommand}"`);
+		const data = await this.Model.connection.executeDbFeature('find', options);
+
+		return {
```

Comment:
Previously this would return `Model` instances by calling a method on the `Model` class.  I figured it would be better to just let the `Model` class do that itself.